docs/internals.md

```diff
@@ -0,0 +1,93 @@
+parameter name (the `Subkey` "value"), as well as its originating `Vset` (the
+`Subkey` "origin"). Since inputs must be dictionaries with a certain format, all
+raw inputs must be initialized using `vflow.helpers.init_args`.
+
+When a `Vset` method such as `fit` or `predict` is called with multiple input
+arguments, the inputs are first combined left to right by Cartesian product; the
+right `Subkey` tuple is always filtered on any matching `Subkeys` in the left
+`Subkey` tuple, and the two keys' remaining `Subkeys` are concatenated. The
+values of the resulting combined dictionary are tuples containing the values of
+the inputs concatenated in the order in which they were passed in. (See
+[`below`](##Subkey) for more info on `Subkey` matching.) Next, the `Vset`
+computes the Cartesian product of the combined input dictionary with
+`Vset.modules` or `Vset.out` (depending on the `Vset` method that was called),
+combining `Subkey` tuples in a similar process to determine which `Vfuncs` to
+apply to which inputs.
+
+If a `Vset` is initialized with `output_matching=True`, then outputs from this
+`Vset` will be matched when used in other `Vsets`. Internally, this means the
+new `Subkey` added to the output dictionary by the `Vset` will have an
+`output_matching` attribute with value `True`, which is used to reject Cartesian
+product combinations when `Subkey` origins do not match or match but their
+values differ. A `Vset` should be initialized with `output_matching=True` when
+it is used multiple times and outputs from that `Vset` need to be combined
+regardless of when it was called, as with a data cleaning `Vset` that is used
+first on a training dataset before model fitting and later on a test dataset for
+model evaluation; this ensures that the data cleaning procedure used in training
+the model matches the data cleaning procedure used when evaluating the model.
+(In contrast, note that in this example we wouldn't need to initialize the
+modeling `Vset` with `output_matching=True`.) On the other hand, if separate
+calls to the `Vset` result in entirely independent outputs, such as is usually
+the case for a `Vset` that does subsampling, then the `Vset` should be
+initialized with `output_matching=False` (the default) to prevent bad matches
+(e.g., unnecessary matching on subsamples of two different datasets).
+
+## Vfunc
+
+When a `Vset` is initialized, the items in the `modules` arg are wrapped in
+`Vfunc` objects, which are like named functions that can optionally support a
+`fit` or `transform` method. If a `Vset` is initialized with `is_async=True`,
+then modules are wrapped in `AsyncModule` objects, which compute asynchronously
+using `ray`. If a `Vset` is initialized with `lazy=True`, `Vset` output
+dictionary values are wrapped in `VfuncPromise` objects, which are lazily
+evaluated. At the moment, `VfuncPromise` objects are only resolved if passed to
+a `Vset` with `lazy=False` downstream or if called manually by the user.
```

Comment:
Similarly here, seems like an easier to read format could be writing a short description and then bullet points like: 

When a `Vset` is initialized, the items in the `modules` arg are wrapped in `Vfunc` objects, which are like named functions that can optionally support a `fit` or `transform` method. 

- `is_async=True`: modules are wrapped in `AsyncModule` objects, which compute asynchronously
using `ray`
- `lazy=True`: `Vset` output dictionary values are wrapped in `VfuncPromise` objects, which are lazily evaluated. At the moment, `VfuncPromise` objects are only resolved if passed to a `Vset` with `lazy=False` downstream or if called manually by the user.